### ISSUE TITLE
[Win10 UAP] Binding Commands to FlyOuts

### ISSUE THREAD

**Issue by Rhywden** [PrismLibrary/Prism#151]
This is probably a limitation of the Win10 UAP framework and I'm sorry if I'm asking this in the wrong place as it is probably something affecting all MVVM frameworks.

Basically, the issue is this:

I have a GridView of items (pupils, for example) which upon clicking them shall show a FlyOut wherein I can make several selections (like: mark absence, homework, ...). This is done by using FlyoutBase.AttachedFlyout in the xaml and FlyoutBase.ShowAttachedFlyout() in the ViewModel's command bound to the Tapped-Event.

The problem becomes this: The DataContext of the FlyOut changes *irrevocably* to the item it's attached to. Furthermore, the outside DataContext (i.e. the ViewModel's) becomes *inaccessible*. Which also means that any and all Commands defined in the ViewModel are non-existant as far as it concerns the FlyOut. At least in xaml, you can't access anything outside the FlyOut and I haven't been able to discover how to access the FlyOut through codebehind in the ViewModel.

Now, I could put stuff in the View's codebehind. Or I could create a hybrid class which contains references to both the item's model and the commands I need. Or I could create a hybrid class and use events instead of commands. Doesn't feel right, though.

Is there some recommended way to deal with this issue? Because there's a dearth of information regarding this.

**Comment by JohnTasler** [comment]
This is definitely more about UWP XAML, and not really at all related to Prism. However, I may have some suggestions if you could send me a small repro project. I just coded up something last week regarding the DataContext of a Flyout for an application I'm working on.

**Comment by briannoyes** [comment]
Yes, I'm afraid there are still a lot of things like this in both Windows 8 WinRT and Windows 10 UWP where the only way to get it done is with either some code in the code behind or sometimes by writing a custom behavior if you need to do it in a lot of places.

Keep in mind that the "no code behind" guidance of MVVM is just that - guidance. There are always scenarios on all of the XAML platforms that you either can't do something without a little code in the code behind at all, or the alternative to a few lines of code in the code behind is a complex workaround that ultimately becomes hard to understand and maintain.

Code that uses the exposed APIs of controls to get the job done in the code behind is View code same as XAML and there is no crime in that. As long as you don't keep sliding down the slippery slope and start putting data manipulation and logic there that is not directly coupled to the UI elements in the View.

**Comment by Rhywden** [comment]
Thanks for the answer. I thought as much but, as I'm still learning about this, also didn't want to reinvent the wheel (badly).

**Comment by DashPriyabrata** [comment]
The similar kind of problem faced by me lately and I fixed by using the following workaround.
1 - As you mentioned the Flyout always takes DataContext of the the control from where it's called which means it no more is bound to the actual ViewModel you are expecting it to attached to.
2 - Since this behaviour of Flyout comes from Framework level, nothing much is under our control. Thus the best solution to this problem would instruct the Flyout to use the actual ViewModel we have intended it to use.
3 - And the best way of doing this is by following this approach:

- Refer the namespace of the viewmodel in XAML header. e.g.- xmlns:vm="using:SampleProject.ViewModels"

- Inside the Flyout set the parent container(Grid/Panel)'s DataContext to the context(ViewModel) you are intending the Flyout to use. For example in my case under the Flyout, "Grid" is the parent container, thus I set DataContext of the grid to the intended ViewModel like this - DataContext="{Binding UI.MainPage, Source={StaticResource DataBindings}}"

- I'm not too sure but I guess the DataContext also can be set in code behind.

That's it. This is how I made it work. May be someone will get benefit from this and may be soon Microsoft will approach this issue we all are facing.
